src/main/kotlin/com/openlattice/data/controllers/DataController.kt

```diff
@@ -0,0 +1,922 @@
+            path = [ENTITY_SET + ENTITY_SET_ID_PATH],
+            method = [RequestMethod.POST],
+            consumes = [MediaType.APPLICATION_JSON_VALUE],
+            produces = [MediaType.APPLICATION_JSON_VALUE, CustomMediaType.TEXT_CSV_VALUE])
+    @Timed
+    fun loadEntitySetData(
+            @PathVariable(ENTITY_SET_ID) entitySetId: UUID,
+            @RequestBody(required = false) selection: EntitySetSelection,
+            @RequestParam(value = FILE_TYPE, required = false) fileType: FileType?,
+            response: HttpServletResponse
+    ): EntitySetData<FullQualifiedName> {
+        setContentDisposition(response, entitySetId.toString(), fileType)
+        setDownloadContentType(response, fileType)
+
+        return loadSelectedEntitySetData(entitySetId, selection, fileType)
+    }
+
+    override fun loadSelectedEntitySetData(
+            entitySetId: UUID, selection: EntitySetSelection, fileType: FileType?
+    ): EntitySetData<FullQualifiedName> {
+        return loadEntitySetData(entitySetId, selection)
+    }
+
+    private fun loadEntitySetData(entitySetId: UUID, selection: EntitySetSelection?): EntitySetData<FullQualifiedName> {
+        if (authz.checkIfHasPermissions(
+                        AclKey(entitySetId),
+                        Principals.getCurrentPrincipals(),
+                        EdmAuthorizationHelper.READ_PERMISSION
+                )) {
+            val entitySet = getEntitySet(entitySetId)
+
+            val entityKeyIds = if (selection == null) Optional.empty() else selection.entityKeyIds
+            val selectedProperties = getSelectedProperties(entitySetId, selection)
+
+            val normalEntitySetIds = if (entitySet.isLinking) entitySet.linkedEntitySets else setOf(entitySetId)
+
+            val authorizedPropertyTypesOfEntitySets = getAuthorizedPropertyTypesForEntitySetRead(
+                    entitySet, normalEntitySetIds, selectedProperties
+            )
+
+            val authorizedPropertyTypes = authorizedPropertyTypesOfEntitySets.values.first()
+            val orderedPropertyNames = LinkedHashSet<String>(authorizedPropertyTypes.size)
+
+            selectedProperties.filter { authorizedPropertyTypes.containsKey(it) }
+                    .map { authorizedPropertyTypes.getValue(it) }
+                    .map { pt -> pt.type.fullQualifiedNameAsString }
+                    .toCollection(orderedPropertyNames)
+
+            return if (entitySet.isLinking) {
+                dgm.getLinkedEntitySetData(
+                        entitySet,
+                        entityKeyIds,
+                        orderedPropertyNames,
+                        authorizedPropertyTypesOfEntitySets
+                )
+            } else {
+                dgm.getEntitySetData(
+                        mapOf(entitySetId to entityKeyIds),
+                        orderedPropertyNames,
+                        authorizedPropertyTypesOfEntitySets
+                )
+            }
+        } else {
```

Comment:
Can you invert this?
So it becomes:
```
if (!currentCase) {
    throw Forbidden()
}
// regular code
```

Reply:
actually I think maybe we should be checking permissions at controller endpoints, not inside of this deeper helper method but I know you just converted this so not a huge deal right now

Reply:
> actually I think maybe we should be checking permissions at controller endpoints, not inside of this deeper helper method but I know you just converted this so not a huge deal right now

Yeah, but all the endpoints reading entity set data come down to this deeper function, so as long as they stay that way, I think it's nicer this way, compared to to duplicating code.


> Can you invert this?
> So it becomes:
> 
> ```
> if (!currentCase) {
>     throw Forbidden()
> }
> // regular code
> ```
Will change it to early return